sw/device/lib/testing/json/mem.c

```diff
@@ -0,0 +1,62 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#define UJSON_SERDE_IMPL 1
+#include "sw/device/lib/testing/json/mem.h"
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/testing/test_framework/ujson_ottf.h"
+
+#define MODULE_ID MAKE_MODULE_ID('j', 's', 'm')
+
+status_t ujcmd_mem_read32(ujson_t *uj) {
+  mem_read32_req_t op;
+  mem_read32_resp_t resp;
+  uintptr_t address;
+  TRY(ujson_deserialize_mem_read32_req_t(uj, &op));
+  if (op.address & (sizeof(uint32_t) - 1)) {
+    return INVALID_ARGUMENT();
+  }
+  address = op.address;
+  resp.value = read_32((void *)address);
+  return RESP_OK(ujson_serialize_mem_read32_resp_t, uj, &resp);
+}
+
+status_t ujcmd_mem_read(ujson_t *uj) {
+  mem_read_req_t op;
+  mem_read_resp_t resp;
+  uintptr_t address;
+  TRY(ujson_deserialize_mem_read_req_t(uj, &op));
+  if (op.data_len > sizeof(resp.data)) {
```

Comment:
Should the read32 also take a data length? The `read_memory32/write_memory32` in opentitanlib for example takes a `[u32]`.

Reply:
No. This is purely for single word reads and writes, to model CSR accesses and avoid excessive I/O for the typical case..